[BoundsSafety] WIP: Make 'counted_by' work for pointer fields; late parsing for 'counted_by' on decl attr position

rapidsna opened the pull request:
This contains a minimum change to make 'counted_by' available for pointer fields and make late parsing work for 'counted_by' on declaration attribute position.

**Here are TODOs:**
- As we promised in [the RFC discussion](https://discourse.llvm.org/t/rfc-enforcing-bounds-safety-in-c-fbounds-safety/70854/68?u=rapidsna), separately guard the late parsing logic under a feature flag (e.g., -fexperimental-late-parse-attributes)
- Make it an error to use 'counted_by' on size-less types (e.g., forward declared structs)

**TODOs that may be done as separate PRs:**
- Fix some test cases that don't work properly for anonymous structs
- Make late parsing work for the type attribute positions
- Have __builtin_dynamic_object_size and UBSan's array-bounds checks take advantage of counted_by on pointer fields (I believe @bwendling has been working on it).
- With -fbounds-safety, make 'counted_by' and the late parsing logic available for more subjects (parameters, variables, return types, etc.)

test-edit

rapidsna commented on clang/test/Sema/attr-counted-by-late-parsed.c:
```diff
@@ -0,0 +1,54 @@
+
+struct at_pointer {
+  int count;
+  struct size_unknown *__counted_by(count) buf; // expected-error{{'counted_by' cannot be applied to an sized type}}
```
TODO: emit an error for 'counted_by' on unknown size like this and update the error message here.

bwendling replied:
This should be fine, because `count` is declared before use&mdash;unless I'm getting type attributes confused with field attributes.. Also, there's a grammar-o in the message: `to _an_ sized type`.

delcypher commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4872,6 +4886,79 @@ void Parser::ParseStructDeclaration(
+
+  ParsedAttributes Attrs(AttrFactory);
+
+  assert(
```
Should this be a diagnostic rather than an assert?

delcypher commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4872,6 +4886,79 @@ void Parser::ParseStructDeclaration(
+
+  // If the Decl is on a function, add function parameters to the scope.
+  {
+    std::unique_ptr<ParseScope> Scope;
```
Are we missing tests for this? This looks like its to support `__counted_by()` on function parameters but I don't think we have that in tests.

delcypher replied:
Actually I guess this isn't testable because of the `  let Subjects = SubjectList<[Field], ErrorDiag>;`?

delcypher commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4997,7 +5087,11 @@ void Parser::ParseStructUnionBody(SourceLocation RecordLoc,
-  MaybeParseGNUAttributes(attrs);
+  MaybeParseGNUAttributes(attrs, &LateFieldAttrs);
+
+  assert(!getLangOpts().CPlusPlus);
```
Should we remove the assert? Are we restricting late parsing to C only?

rapidsna replied:
@delcypher C++ already has late parsing support for C++ classes and structs and I believe it's handled separately in `ParseDeclCXX.cpp`. I added the assertion to ensure that the code path isn't taken in C++. We should run clang tests if that breaks anything.

bwendling replied:
Is it possible to combine C++'s late parsing with your implementation? Or is there little overlap?

delcypher commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1599,6 +1599,13 @@ defm double_square_bracket_attributes : BoolFOption<"double-square-bracket-attri
   LangOpts<"DoubleSquareBracketAttributes">, DefaultTrue, PosFlag<SetTrue>,
  NegFlag<SetFalse>>;
 
+defm experimental_late_parse_attributes : BoolFOption<"experimental-late-parse-attributes",
```
@rapidsna FYI. Here's a sketch of an experimental late parsing flag. It's not quite done yet and I think it's big enough that it deserves it's own PR. Just pushing it here to give you an idea of my plans

bwendling commented on clang/test/Sema/attr-counted-by-late-parsed.c:
FYI: I'm planning on having a separate `test/Sema/attr-counted-by-for-pointers.c` test file for pointers and rename the current file something like `test/Sema/attr-counted-by-for-flexible-array-members.c`.

rapidsna replied:
> This should be fine, because count is declared before use—unless I'm getting type attributes confused with field attributes..

`struct size_unknown *__counted_by(count) buf` indicates that `buf` has `sizeof(struct size_unknown) * count` bytes. So the problem is here that the compiler doesn't know what's the actual byte size of `buf` because `sizeof(struct unknown_size)` is not available here.

delcypher commented:
@rapidsna @bwendling I've put the experimental feature flag for late parsing in a separate PR (#88596). I've not updated this PR to be based on that PR yet but will do soon. I figured the code for experimental feature flag was large enough that it deserved its own PR to make things easier to review.

delcypher commented:
The implementation of the experimental late attribute changed (#88596) so I've rebased this PR on it and fixed it up to work as intended.

delcypher commented:
@rapidsna I've addressed the "Make it an error to use 'counted_by' on size-less types (e.g., forward declared structs)" TODO.

I still need to test this PR properly and write a proper commit message explaining what this does.

rapidsna commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -8547,15 +8547,25 @@ static const RecordDecl *GetEnclosingNamedOrTopAnonRecord(const FieldDecl *FD) {
 
+  const auto FieldTy = FD->getType();
+  if (FieldTy->isPointerType() &&
+      FieldTy->getPointeeType()->isIncompleteType()) {
```
`isIncompleteType()` won't be sufficient. We should also prevent a pointer to sizeless type, function type, and struct with flexible array member, etc.

delcypher committed:
[BoundsSafety] Allow 'counted_by' attribute on pointers in structs in C

Previously the attribute was only allowed on flexible array members.
This patch patch changes this to also allow the attribute on pointer
fields in structs and also allows late parsing of the attribute in some
contexts.

For example this previously wasn't allowed:

```
struct BufferTypeDeclAttributePosition {
  size_t count;
  char* buffer __counted_by(count); // Now allowed
}
```

Note the attribute is prevented on pointee types where the size isn't
known at compile time. In particular pointee types that are:

* Incomplete (e.g. `void`) and sizeless types
* Function types (e.g. the pointee of a function pointer)
* Struct types with a flexible array member

This patch also introduces late parsing of the attribute when used
in the declaration attribute position. For example

```
struct BufferTypeDeclAttributePosition {
  char* buffer __counted_by(count); // Now allowed
  size_t count;
}
```

is now allowed but **only** when passing
`-fexperimental-late-parse-attributes`.  The motivation for using late
parsing here is to avoid breaking the data layout of structs in existing
code that want to use the `counted_by` attribute. This patch is the
first use of `LateAttrParseExperimentalExt` in `Attr.td` that was
introduced in a previous patch.

Note by allowing the attribute on struct member pointers this now allows
the possiblity of writing the attribute in the type attribute position.
For example:

```
struct BufferTypeAttributePosition {
  size_t count;
  char *__counted_by(count) buffer; // Now allowed
}
```

However, the attribute in this position is still currently parsed
immediately rather than late parsed. So this will not parse currently:

```
struct BufferTypeAttributePosition {
  char *__counted_by(count) buffer; // Fails to parse
  size_t count;
}
```

The intention is to lift this restriction in future patches. It has not
been done in this patch to keep this size of this commit small.

There are also several other follow up changes that will need to be
addressed in future patches:

* Make late parsing working with anonymous structs (see
  `on_pointer_anon_buf` in `attr-counted-by-late-parsed-struct-ptrs.c`).
* Allow `counted_by` on more subjedts (e.g. parameters, returns types)
  when `-fbounds-safety` is enabled.
* Make use of the attribute on pointer types in code gen (e.g.
for `_builtin_dynamic_object_size` and UBSan's array-bounds checks).

This work is heavily based on a patch originally written by Yeoul Na.

rdar://125400257

delcypher commented:
@rapidsna I've put my version of this PR in my own pull request because when I edit the PR summary here it looks like you wrote it which is weird: https://github.com/llvm/llvm-project/pull/90786

delcypher commented:
@rapidsna You may want to close this PR but I'll leave this up to you.
